presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestIntermediateAggregationOptimizer.java

```diff
@@ -0,0 +1,163 @@
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.FINAL;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.INTERMEDIATE;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.PARTIAL;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static com.facebook.presto.tpch.TpchMetadata.TINY_SCHEMA_NAME;
+
+public class TestIntermediateAggregationOptimizer
+{
+    private final LocalQueryRunner queryRunner;
+
+    public TestIntermediateAggregationOptimizer()
+    {
+        Session defaultSession = testSessionBuilder()
+                .setCatalog("local")
+                .setSchema(TINY_SCHEMA_NAME)
+                .setSystemProperty(SystemSessionProperties.ENABLE_INTERMEDIATE_AGGREGATIONS, "true")
+                .build();
+
+        this.queryRunner = new LocalQueryRunner(defaultSession);
+        queryRunner.createCatalog(queryRunner.getDefaultSession().getCatalog().get(),
+                new TpchConnectorFactory(1),
+                ImmutableMap.of());
+    }
+
+    @Test
+    public void testIntermediate()
+    {
+        ExpectedValueProvider<FunctionCall> rawInputCount = PlanMatchPattern.functionCall("count", false, ImmutableList.of());
+        ExpectedValueProvider<FunctionCall> partialInputCount = PlanMatchPattern.functionCall("count", false, ImmutableList.of(anySymbol()));
+
+        assertPlan("SELECT COUNT(*) FROM orders",
+                anyTree(
+                        aggregation(
+                                ImmutableList.of(ImmutableList.of()),
+                                ImmutableMap.of(Optional.empty(), partialInputCount),
+                                ImmutableMap.of(),
+                                Optional.empty(),
+                                FINAL,
+                                exchange(LOCAL, GATHER,
+                                        exchange(REMOTE, GATHER,
+                                                aggregation(
+                                                        ImmutableList.of(ImmutableList.of()),
+                                                        ImmutableMap.of(Optional.empty(), partialInputCount),
+                                                        ImmutableMap.of(),
+                                                        Optional.empty(),
+                                                        INTERMEDIATE,
```

Comment:
INTERMEDIATE takes partial and outputs partial,
PARTIAL takes raw and outputs partial